### ISSUE TITLE
Not able to browse Documents folder in Mac

### ISSUE THREAD

**Issue by qwertangel** [zaproxy/zaproxy#6444]
**Describe the bug**
Mac mojave and catalina has new file permission model to improve security (aka TCC). Applications now need to ask for permissions to access Documents/Downloads/Desktop folders.   I installed (and reinstalled multiple times) OWasp Zap in mac catalina 10.15.7. It starts up fine, however when trying to open a previously stored session (from another laptop) in my documents folder, I am not able to see any files/folder inside Documents folder. This is same for Desktop and downloads folder. (Presumably because of Mac's TCC permissions.). I tried resetting TCC for org.zaproxy.zap.ZAP and gave 'full disk' permissions to Zap. However even after that Zap is not able to see any files/folders in above folders.

**To Reproduce**
1. Open Zap
2. Click 'File' menu
3. Click 'Open Session'
4. Set file selector to 'All files.
5. Browse to Documents folder
6. Observe that no files/folders are visible.
7. If browseing to any other folder other than Documents/Downloads/Desktop folders, all files and folders are visible normally.

**Expected behavior**
Files/folder should be visible.

**Screenshots**
![image](https://user-images.githubusercontent.com/3331518/107417382-6081f600-6aca-11eb-80f1-f3eb444ffe6f.png)

**Software versions**
 - ZAP: 2.10.0
 - Add-on: None
 - OS: MacOs Catalina 10.15.7 (19H2)
 - Java: Zap inbuilt jdk
 - Browser: Not applicable

**Errors from the zap.log file**
No errors in log file related to file dialog.

**Additional context**
1) Tried reinstalling Zap multiple times. Zap never asked me for tcc permissions
2) Gave full disk permissions to Zap in Mac Privacy settings.
2) Reset tcc permissions using 'sudo tccutil reset All org.zaproxy.zap.ZAP'
![image](https://user-images.githubusercontent.com/3331518/107417736-be164280-6aca-11eb-92d5-d465f40209a6.png)

**Would you like to help fix this issue?**
Can provide any logs or debugging steps to reproduce the issue

**Comment by thc202** [comment]
This would be a configuration problem in macOS not ZAP.

Re the ideal behaviour, does the JRE provide APIs for that type of integration?

**Comment by qwertangel** [comment]
I think it is not just macOS config problem since MacOS does provide options (other apps can have their permissions removed/added using UI/command line). However Zap doesn't seem to be compliant since it doesn't ask for permissions and neither standard MacOS ways to add/remove permissions seem to work.

**Comment by thc202** [comment]
Are you referring to other Java applications or "native"? If Java/JRE does not provide APIs to ask for permissions I don't think we'll be able provide that feature.

**Comment by qwertangel** [comment]
I am using macos native package (which seems to using java under the hoods but I could be wrong).

**Comment by thc202** [comment]
I was asking about the other applications that ask for permissions. (Yes, ZAP is using Java, that's why I was saying that we would need APIs to interact with macOS.)

**Comment by psiinon** [comment]
Interestingly enough I can reproduce this on my Mac for some directories but not others.
I'm sure its a MacOS configuration issue not a ZAP problem, but I'm not an experienced enough Mac user to know the correct solution :/
It would be really good to get a definitive set of instructions on how to fix this on Macs which we can include somewhere, maybe the FAQs. 
Maybe leave this issue open until we have it?

**Comment by garthoid** [comment]
I am encountering this on Catalina 10.15.7.

In addition I have encountered a few "you do not have permission" when I attempt to save in a directory I own as the user running the application as.

**Comment by blesson3** [comment]
I ran into the same issue as @carefulcomputer with the latest release (`v2.10.0` - Dec 17, 2020), but when I built the latest code (`7017aea`) from source the problem was fixed (I could see visit and see all folders/files). Must have been fixed in a commit between the latest release and now.

I also had given full-disk permissions to OWASP ZAP.app through the standard `Preferences -> Security & Privacy -> Full Disk Permissions` UI. This setting has no effect on this bug/misconfig for the latest release (`v2.10.0`) and the latest built source.

I'm running macOS Big Sur (11.1)

```
$ sw_vers
ProductName:	macOS
ProductVersion:	11.1
BuildVersion:	20C69
```

**Comment by psiinon** [comment]
I'm not aware of any change we've made to ZAP which could have affected this, so I still think its a MacOS configuration issue.
Dont forget that when you run ZAP after you have built it locally then you may well be picking up the permissions from either Java or your IDE.

**Comment by psiinon** [comment]
> ZAP provides its own Java binary: `/Applications/OWASP ZAP.app/Contents/Home/bin/java`
> 
> If that one's being used, it's that process which should query MacOS for access permissions to ~/Documents. And since it's provided as part of ZAP, we should find a way of making ZAP trigger the query to MacOS.
> 
> Here is an issue on another Github project that clearly outlines the situation @thc202 -> [tofi86/universalJavaApplicationStub#85](https://github.com/tofi86/universalJavaApplicationStub/issues/85)

Thanks for this - that looks very useful!

**Comment by tsluyter** [comment]
> so I still think its a MacOS configuration issue.

Yes and no :) The fact that MacOS is blocking ZAP from accessing the ~/Documents folder is a MacOS config issue, correct. But for most users, the best/easiest way to have that configuration occur is to have the application prompt the OS with an access request.

**Comment by blesson3** [comment]
@psiinon Can confirm that the locally built versions don't have this bug. I built the latest code (`7017aea`) and at the `v2.10.0` release tag, and I can access all folders/files in ZAP through the dev build. Even though I cannot access some folders/files through the `v2.10.0` released version. I updated my above comment to reflect this.

I'm thinking that the access permissions for dev builds are inherited from ide or system java (as @psiinon said). ZAP should still support the macOS-specific query to request full-disk (or granular folder) access permissions.

**Comment by kmcquade** [comment]
I also ran into this issue when I tried to save Context. I saw the [the comment](https://github.com/zaproxy/zaproxy/issues/6444#issuecomment-799284733) by @tsluyter above and tried it, but that didn't work either. Screenshot below.

![image](https://user-images.githubusercontent.com/3422255/116273176-8759db00-a74f-11eb-9218-0776535bfd3a.png)

**Comment by tsluyter** [comment]
@kmcquade did you allow ZAP access to your ~/Library/ directory, or to ~/Documents/? Because by default ZAP exports to ~/Library/........... as your save-dialog shows.

**Comment by kmcquade** [comment]
@tsluyter - I granted "Full Disk Access", as you noted in the edit to your comment. Presumably that should work for both ~/Library/ and ~/Documents but it didn't work for either.

**Comment by carefulcomputer** [comment]
Since this issue is confirmed, should we mark it as a 'bug' instead of 'enhancement' ?

**Comment by psiinon** [comment]
Is it really a bug in ZAP or just missing / incomplete documentation?
As far as I can tell its Mac's being a pain, not something we can actually fix in the ZAP source code, but I'd be v happy to be proved wrong...

**Comment by carefulcomputer** [comment]
I agree Mac can be pain in neck. However, Mac does provides ability to ask for permissions (other apps work similar way) and if Zap is not using it, then I would say issue is with Zap.

**Comment by psiinon** [comment]
Links please! If someone can explain what we should be doing then we can definitely look into doing it.
Otherwise I propose keeping this as an enhancement request in the hope that it annoys someone else enough for them to give us some clues :P
FYI I have added both Java and ZAP to the list of apps with access to:

- Full Disk Access
- Files and Folders

I can save contexts to my `~/Library/Application Support/ZAP/contexts` directory.
Can anyone whos had this issue try that and report back here?

**Comment by gromanas** [comment]
Had the same issue. Just change the name of a context from a `url.context` (e.g. https://github.com.context) to another name without using a url (e.g. github.context).

**Comment by ajit-y** [comment]
Hi, I encountered the same problem today with ZAP on my mac (OS - Monterey) that ZAP did not have access to any content in Documents folder even after I gave Full disk access to it.

While searching for the solution I came across this thread. Maybe people already figured it out but still posting here what worked for me in case it is of any help to anyone.

Workaround:- I created a new directory under my home directory i.e. /Users/home/ZAP-Sessions , ZAP was able to access and save sessions there. So I just ended up using this ZAP-Sessions directory for saving and opening sessions.

**Comment by n8henrie** [comment]
Same issues. "Full Access" didn't help.

Because I've given `Terminal.app` access, I'm able to access `~/Downloads` when I start it from the CLI as so:

```console
$ /Applications/OWASP\ ZAP.app/Contents/Java/zap.sh
```

**Comment by vitobotta** [comment]
> Same issues. "Full Access" didn't help.
> 
> Because I've given `Terminal.app` access, I'm able to access `~/Downloads` when I start it from the CLI as so:
> 
> ```
> $ /Applications/OWASP\ ZAP.app/Contents/Java/zap.sh
> ```

This is pretty annoying but at least it works, thanks.  I added an alias to my .zshrc

```bash
alias zap="/Applications/OWASP\ ZAP.app/Contents/Java/zap.sh 2>&1 > /dev/null  &"
```

So I just type `zap` in the terminal and am not bothered with the ouput.
